pkg/aggregator/aggregator.go

```diff
@@ -0,0 +1,226 @@
+package aggregator
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/kit/log"
+
+	"github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb/chunkenc"
+)
+
+// Aggregator implement storage.Aggregator and runs a set of PromQL queries
+// on the in-memory batch before committing to the underlying appendable.
+type Aggregator struct {
+	logger     log.Logger
+	appendable storage.Appendable
+	engine     *promql.Engine
+	rules      []Rule
+}
+
+// Rule for scrape-time aggregation.
+type Rule struct {
+	Name   string            `yaml:"name"`
+	Expr   string            `yaml:"expr"`
+	Labels map[string]string `yaml:"labels,omitempty"`
+}
+
+// New makes a fresh Aggregator.
+func New(logger log.Logger, appendable storage.Appendable, rules []Rule) *Aggregator {
+	engine := promql.NewEngine(promql.EngineOpts{
+		Logger:        logger,
+		MaxSamples:    1000000,
+		Timeout:       time.Minute,
+		LookbackDelta: 15 * time.Minute,
+	})
+
+	return &Aggregator{
+		logger:     logger,
+		appendable: appendable,
+		engine:     engine,
+		rules:      rules,
+	}
+}
+
+// Appender implements storage.Appendable.
+func (a *Aggregator) Appender(ctx context.Context) storage.Appender {
+	return &batch{
+		ctx:        ctx,
+		aggregator: a,
+		appender:   a.appendable.Appender(ctx),
+	}
+}
+
+type batch struct {
+	ctx        context.Context
+	aggregator *Aggregator
+	appender   storage.Appender
+	samples    []sample
+	ts         int64
+}
+
+func (b *batch) Add(l labels.Labels, t int64, v float64) (uint64, error) {
+	b.samples = append(b.samples, sample{l, t, v})
+	return b.appender.Add(l, t, v)
+}
+
+func (b *batch) AddFast(ref uint64, t int64, v float64) error {
+	return b.appender.AddFast(ref, t, v)
+}
+
+func (b *batch) Commit() error {
+	for _, r := range b.aggregator.rules {
+		if err := b.execute(r); err != nil {
+			return err
+		}
+	}
+
+	return b.appender.Commit()
+}
```

Comment:
Very cool. If I understand correctly, this is going to commit both aggregated and non-aggregated samples to the WAL, right? 

I think this will end up with a usage pattern of the aggregation rules always being combined with metric_relabeling_rules to drop the original non-aggregated samples. 

It would be cool if we could defer writes to the underlying appender and only write aggregated samples and samples that weren't part of an aggregation rule, but I fear that might be more complicated than it's worth.